### DISCUSSION TITLE
Can we expose Java objects as records?

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#6614] [Ideas]
Imagine that we have:

```scala
let reader: BufferedReader = ...
```

We would like to write: 

```scala
let b = reader.read();
reader.close()
```

What would happen if we gave `reader` the record type?:

```scala
{
  close = Unit -> Unit \ IO,
  lines = Unit -> Stream<String> \ IO,
  mark = Int-> Unit \ IO,
  markSupported = Unit -> Bool \ IO,
  read = Unit -> Int32 \ IO,
  readLine = Unit -> String \ IO,
  ready = Unit -> Bool \ IO,
  reset = Unit -> Unit \ IO,
  skip = Int64 -> Int64 \ IO,
}
```

Why?

- We get type inference.
- We can use the "dot" as we are used to.
- We can use objects as records with structural typing.
- We get auto-complete for free.

But there are challenges and problems:

- [ ] We would have to creature closures for each method (research problem: how to optimize?)
- [ ] We should expect type inference to be slow/costly for many-field objects (research problem: what can be done?)
- [ ] We cannot readily support overloading (oof). In fact there are two `read` methods on `BufferedReader`.
- [ ] Debugging could be a bit of a nightmare because objects lose their name and just become a bunch of fields.

Did I miss any?

Of these overloading is the foremost and immediate problem (but the rest are still very serious). 

A few solutions:

- We simply have fields `read_1` and `read_3` using the field arity. (Note this is not enough because methods can overload based on types). 
- We use scoped fields (this is very un-ergonomic and what would be the order?)
- We have one field which contains _another record_ with all the methods. (Ugh).
- Programmers have to supply annotations (ugh...)

**Comment by magnus-madsen** [comment]
@stephentetley @mlutze @JonathanStarup @AStenbaek @jaschdoc @sockmaster27 I would appreciate your brainstorming power on this. 

I would like interop to be more accessible and useful. I think it is key to attract more programmers.

**Comment by mlutze** [comment]
Some more problems:
- how to deal with fields
- how to deal with private members/incapsulation/internal invariants
- how to deal with instanceof checks
  - If our records extend the appropriate class, what happens with final classes?

IMO we may have that a BufferedReader is a record, but we don't have that a record is a BufferedReader.

**Reply by magnus-madsen** [reply]
> IMO we may have that a BufferedReader is a record, but we don't have that a record is a BufferedReader.

I think this is a greater problem: How would you pass back a `BufferedReader` where it is expected? One option would be for the record to carry a "ghost field" with its type.

**Comment by magnus-madsen** [comment]
We went in a more normal direction :)
